src/util/green_reordering.cpp

```diff
@@ -0,0 +1,217 @@
+#include <util/green_reordering.h>
+#include <array>
+
+bool expr_variable_reordering::run(expr2tc &e)
+{
+  switch(e->expr_id)
+  {
+    // BINARY OPS
+  case expr2t::expr_ids::add_id:
+  case expr2t::expr_ids::mul_id:
+    run_on_binop(e);
+    break;
+    // RELATIONS
+  case expr2t::expr_ids::equality_id:
+  case expr2t::expr_ids::notequal_id:
+  case expr2t::expr_ids::lessthan_id:
+  case expr2t::expr_ids::greaterthan_id:
+  case expr2t::expr_ids::lessthanequal_id:
+  case expr2t::expr_ids::greaterthanequal_id:
+    run_on_relation(e);
+    break;
+    // Negations
+  case expr2t::expr_ids::neg_id:
+  case expr2t::expr_ids::not_id:
+    run_on_negation(e);
+    break;
+  default:
+    break; // don't care
+  }
+  return true;
+}
+
+void expr_variable_reordering::run_on_binop(expr2tc &expr)
+{
+  std::shared_ptr<arith_2ops> arith;
+  arith = std::dynamic_pointer_cast<arith_2ops>(expr);
+  // If arith is false then the cast failed
+  assert(arith);
+
+  /** Reorder both sides recursively
+   *
+   * The idea of this algorithm is to parse all symbols and values of a
+   * expression with the same operator and rewrite it.
+   *
+   * 1. Create a list of symbols and values
+   * 2. Order it in crescent order
+   * 3. Change all values
+   */
+
+  run(arith->side_1);
+  run(arith->side_2);
+
+  symbols_vec symbols; // annotates all inner symbols
+  values_vec values;   // annotates all inner values
+
+  // 1. Create a list of symbols and values
+  this->transverse_read_binop(arith, symbols, values);
+
+  // 2. Order it in crescent order
+  sort(symbols.begin(), symbols.end(), [](const auto &lhs, const auto &rhs) {
+    return lhs->get_symbol_name() > rhs->get_symbol_name();
+  });
+
+  // A constant propagation should have been executed prior to this
+  assert(values.size() <= 1);
+
+  // 3. Change all values
+  this->transverse_replace_binop(arith, symbols, values);
+}
+
+void expr_variable_reordering::run_on_negation(expr2tc &expr)
+{
+  if(expr->expr_id == expr2t::expr_ids::neg_id)
+  {
+    std::shared_ptr<arith_1op> arith;
```

Comment:
why are you using the `std::shared_ptr<arith_1op>` directly? Can you replace this and all the similar objects with `expr2tc`?